oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java

```diff
@@ -0,0 +1,278 @@
+
+        IndexingProgressReporter progressReporterPerTask =
+                new IndexingProgressReporter(IndexUpdateCallback.NOOP, NodeTraversalCallback.NOOP);
+        progressReporterPerTask.setMessagePrefix("Dumping from " + TRAVERSER_ID_PREFIX);
+
+        this.filter = (id) -> {
+            try {
+                progressReporterPerTask.traversedNode(() -> id);
+            } catch (CommitFailedException e) {
+                throw new RuntimeException(e);
+            }
+            traversalLog.trace(id);
+            return true;
+        };
+
+        // Default retries for 5 minutes.
+        this.retryDuringSeconds = ConfigHelper.getSystemPropertyAsInt(
+                OAK_INDEXER_PIPELINED_MONGO_CONNECTION_RETRY_SECONDS,
+                DEFAULT_OAK_INDEXER_PIPELINED_MONGO_CONNECTION_RETRY_SECONDS);
+        this.dbCollection = MongoDocumentStoreHelper.getDBCollection(mongoStore, collection);
+
+        //TODO This may lead to reads being routed to secondary depending on MongoURI
+        //So caller must ensure that its safe to read from secondary
+//        this.readPreference = MongoDocumentStoreHelper.getConfiguredReadPreference(mongoStore, collection);
+        this.readPreference = ReadPreference.secondaryPreferred();
+        LOG.info("Using read preference {}", readPreference.getName());
+
+    }
+
+    @Override
+    public Result call() throws Exception {
+        String originalName = Thread.currentThread().getName();
+        Thread.currentThread().setName("mongo-dump");
+        try {
+            LOG.info("Starting to download from MongoDB.");
+            MongoCollection<BasicDBObject> dbCollection = MongoDocumentStoreHelper.getDBCollection(mongoStore, collection);
+            //TODO This may lead to reads being routed to secondary depending on MongoURI
+            //So caller must ensure that its safe to read from secondary
+
+            this.nextLastModified = getFirstLastModified(dbCollection);
+            long upperBoundLastModified = getLatestLastModified(dbCollection) + 1;
+            this.lastIdDownloaded = null;
+            Instant failureTimestamp = null;
+            long retryInterval = retryInitialIntervalMillis;
+
+            Map<String, Integer> exceptions = new HashMap<>();
+            while (true) {
+                try {
+                    if (lastIdDownloaded != null) {
+                        LOG.info("Recovering from broken connection, finishing downloading documents with _modified={}", nextLastModified);
+                        downloadRange(new TraversingRange(new LastModifiedRange(nextLastModified, nextLastModified + 1), lastIdDownloaded));
+                        // Reset the retries
+                        failureTimestamp = null;
+                        // Continue downloading everything starting from the next _lastmodified value
+                        downloadRange(new TraversingRange(new LastModifiedRange(nextLastModified + 1, upperBoundLastModified), null));
```

Comment:
I think it cannot happen. `upperBoundLastModified` is set to 1 plus the highest _modified value that can be returned by the traversal queries to MongoDB:
```
long upperBoundLastModified = getLatestLastModified(dbCollection) + 1;
```
And `nextLastModified` is only assigned to values returned by the query:
```
this.nextLastModified = next.getLong(NodeDocument.MODIFIED_IN_SECS);
```
Furthermore, the traversing query is exclusive of the upper bound, see `TraversingRange.getFindQuery()`, so it will always return `_modfied` values lower than `upperBoundLastModfied`.

So at most, `nextLastModifed+1` will be equal to `upperBoundLastModified`. Or is there something wrong with my reasoning?